### ISSUE TITLE
KDE4 libraries duplication/inconsistency

### ISSUE THREAD

**Issue by ambrop72** [NixOS/nixpkgs#6883]
The recent "KDE Applications" (`pkgs.kdeApps_stable`) releases come with a release of KDE4 kdelibs. This kdelibs identifies itself as version 4.14.6, as can be seen in applications within this package set that end up linked to it (About dialog, ldd). However, we also have kdelibs and other kde4 libraries in the original KDE4 packages (`pkgs.kde4`). The latest kdelibs version there is 4.14.3.

This by itself is not too bad, but the problem is that the `kdeApps` expressions make use of certain packages from `pkgs.kde4` as dependencies. This is not good, because an application may then depend on both variants of kdelibs. Which one ends up being loaded at runtime may be a matter of chance more than anything else.

This can be checked using `ldd`. In my case, for example, Akregator (in `pkgs.kdeApps_stable.kdepim`) is linked to the new kdelibs, while Okular (`pkgs.kdeApps_stable.okular`) is linked to the old kdelibs.

```
$ ldd $(nix-build -A pkgs.kdeApps_stable.kdepim --no-out-link)/bin/akregator |grep kdecore
        libkdecore.so.5 => /nix/store/safajj2kzs57bx6y4mc025882wm0sf88-kdelibs-4.14.6/lib/libkdecore.so.5 (0x00007f39d75a7000)
$ ldd $(nix-build -A pkgs.kdeApps_stable.okular --no-out-link)/bin/okular |grep kdecore
        libkdecore.so.5 => /nix/store/2rhs2i291j6bz4vrdpppvvhqr8v86laa-kdelibs-4.14.3/lib/libkdecore.so.5 (0x00007f3cd3acf000)
```

Any suggestions we could clean this up? I see two ways:
1. Get rid of dependencies on `kde4` within the kdeApps expression. Instead, somehow build the KDE4 dependencies that are not part of kdeApps within the kdeApps expression, so they are built agains the kdeApps kdelibs.
2. Do not build kdelibs as part of the kdeApps expression, but rely on the expressions in kde4. Preferably we modify those so that the updated kdelibs are used, and the older 4.14.3 would be gone.

I have made a proof of concept of the first approach: https://github.com/ambrop72/nixpkgs/commit/8d057da664a0dd87cc4955c83a84dee96069bd75

I have noticed that the of the `kdeApps` KDE4 applications, the ones that are linked to the new 4.14.6 kdelibs will have some problems with the application styles (the KDE5 style settings have no effect on them). But this seems like some simple integration that is done in kde4 kdelibs but not in kdeApps kdelibs.

**Comment by ttuegel** [comment]
As the `kdeApps` maintainer, I strongly oppose your second suggestion. The `kde4` expressions have not been actively maintained in more than a year. The correct thing to do is remove the references to `kde4` from `kdeApps`. I have not done this yet because it will happen automatically over the next upstream release cycle, as upstream migrates everything to KDE Frameworks 5. (Actually, it looks like it may take two upstream releases.)

I cannot reproduce the application styles issue you mentioned. Can you open a separate issue for us to discuss that?

**Comment by ambrop72** [comment]
@ttuegel I do not understand your preference. From my perspective, this is the easiest solution to "correctly" support KDE4 code in kdeApps. As you can see my pull request removes more code than it adds :)

- Since as you have mentioned, KDE4 code will be phased out of kdeApps, there shouldn't be much KDE4 related maintenance other than updating the kde4.kdelibs version when a new version is released, and removing all kde4 dependencies in kdeApps as KDE4 is phased out.
- Manual changes to the auto-generated KDE4 manifest are fine because there presumably won't be another full KDE4 release - they aren't a maintenance burden.
- KDE4 apps in kdeApps get to rely on the stable kde4 expressions for their dependencies, instead of having to rewrite them.
- I don't think we should worry about breaking any KDE4 apps with the kdelibs updates, after all kdelibs updates should be fully backward compatible.

**Comment by ambrop72** [comment]
As far as the style problem goes, I'm seeing it when running kdeApps Akregator and Okular in Plasma5. So in systemPackages, I have `kde5.kdepim kde5.okular`. Okular looks nice with the Breeze theme, but Akregator (which gets linked to the kdeApps kdelibs) is ugly with some default theme. You might want to make sure you're actually running the kdeApps versions of both these, not kde4.

**Comment by ttuegel** [comment]
<blockquote>
Since as you have mentioned, KDE4 code will be phased out of kdeApps, there shouldn't be much KDE4 related maintenance other than updating the kde4.kdelibs version when a new version is released, and removing all kde4 dependencies in kdeApps as KDE4 is phased out.
</blockquote>
The KDE4 dependencies are going away, but the packages are not; they are just being replaced by their KDE5 versions as they become available. You are suggesting that we manually remove all the KDE4 packages from `kdeApps`, then manually add them all back when they become available in KDE5. Meanwhile, the KDE4 versions are receiving bugfixes, so we would need to be updating the `kde4` versions.

<blockquote>
Manual changes to the auto-generated KDE4 manifest are fine because there presumably won't be another full KDE4 release - they aren't a maintenance burden.
</blockquote>
They are a maintenance burden, because someone has to do them.

<blockquote>
KDE4 apps in kdeApps get to rely on the stable kde4 expressions for their dependencies, instead of having to rewrite them.
</blockquote>
I don't understand this comment. Who is rewriting what?

<blockquote>
I don't think we should worry about breaking any KDE4 apps with the kdelibs updates, after all kdelibs updates should be fully backward compatible.
</blockquote>
Yes, I agree updating kdelibs won't break anything. As we speak, I am updating `kde4` so there will only be one version of kdelibs on the system. But, I don't need to (and I'm not going to) remove anything from `kdeApps` to do this.

I also suspect that updating kdelibs everywhere will fix the style problem. KDE and Qt are extremely sensitive to having multiple versions around. This may be why you see this issue and I don't, but I won't know until everything finishes building.

**Comment by ambrop72** [comment]
> You are suggesting that we manually remove all the KDE4 packages from kdeApps, then manually add them all back when they become available in KDE5. 

Nah, that was just an idea. I think it's all right to just do it for kdelibs. And maybe some other library in the future if it solves a problem. I'm fine with having KDE4 apps in kdeApps while they're being released.

> I don't understand this comment. Who is rewriting what?
This is in relation to my Option 1, manually writing derivations for KDE4 dependencies of kdeApps KDE4 apps that are not released as part of kdeApps. See what I had to do in my attempt of this option, https://github.com/ambrop72/nixpkgs/commit/8d057da664a0dd87cc4955c83a84dee96069bd75 .

> Yes, I agree updating kdelibs won't break anything. As we speak, I am updating kde4 so there will only be one version of kdelibs on the system. But, I don't need to (and I'm not going to) remove anything from kdeApps to do this.

Ah, we are on the same page then :) I already made a pull request, in case you missed it. https://github.com/NixOS/nixpkgs/pull/6904

**Comment by ambrop72** [comment]
Oh, I mean, I would do it by removing *just* kdelibs from kdeApps and patch it in kde4, as in my pull request. It seems safer to me, than removing kdelibs from kde4 and making it rely on kdeApps.kdelibs. But I don't have a strong opinion on this if it works in the end.

**Comment by ambrop72** [comment]
I foresee a problem with making kde4 rely on kdeApps.kdelibs, if we have multiple versions of kdeApps (latest, stable). When one uses the newer version but kde4 is using the older version, this problem will recur.

**Comment by ttuegel** [comment]
<blockquote>
I foresee a problem with making kde4 rely on kdeApps.kdelibs, if we have multiple versions of kdeApps (latest, stable). When one uses the newer version but kde4 is using the older version, this problem will recur.
</blockquote>
This is already a problem for everything that uses Qt. In the KDE Frameworks packages, I solve it by using `override` wherever I call a Qt package from outside the collection. I can do the same here by overriding `kde4` wherever needed.

**Comment by ambrop72** [comment]
@ttuegel I'm not quite sure what you're planning to do. You said you'll update kdelibs in kde4. Are you planning to update it but leave kdeApps.kdelibs as is? Because that makes no sense to me. If we want to have the same kdelibs version for everything it's most sensible to update kde4.kdelibs and "drop" kdelibs from kdeApps expressions. After all we're already depending on kde4.kdelibs via things like kde4.akonadi and kde4.baloo, there is no reason to bring in another build of kdelibs, it will only cause problems.

I'm not familiar with override, but I think just dropping kdeApps.kdelibs solves the problem at its root rather than fixing up a broken design.

**Comment by ttuegel** [comment]
@ambrop72 I want to be able to have multiple versions of `kdelibs` in `nixpkgs` at one time so I can have `kdeApps_latest != kdeApps_stable`. If I can only have one `kdelibs` build, I either have to upgrade `kdeApps_stable` to the latest `kdelibs` when it comes out (defeating the purpose of having a stable collection) or I have to re-test everything when I move things from `kdeApps_latest` to `kdeApps_stable` (defeating the purpose of having `kdeApps_latest`). The important point is that everything in `kdeApps_stable` use only one `kdelibs` version, and likewise everything in `kdeApps_latest`.

**Comment by ambrop72** [comment]
Well, not quite defeating the purpose of stable kdeApps - it still works for everything except kdelibs. I don't see having a single kdelibs version as a problem since kdelibs is mature software presumably only receiving bug fixes with minor releases.

**Comment by ambrop72** [comment]
But, I imagine it should be easy to parametrize kde4 to use a specific kdelibs version. Basically a parameter that overrides the kdelibs version and source URL. We'd still drop kdeApps.kdelibs, but via this override, the correct kdeApps kdelibs version would still be used. Even more, it would also be used for kdeApps dependencies in kde4 (akonadi, ballo). Do you think this is okay?

**Comment by ttuegel** [comment]
@ambrop72 Well, that's almost exactly what I suggested, so yes, that sounds good! I was just about to upload it after a few tests finish. I have `kde4` take a `kdelibs` parameter that is a complete `kdelibs` build, rather than version + source URL (in Nixpkgs we normally pass around derivations--built packages--rather than strings and things). I could go further and hide `kdeApps.kdelibs`, but I don't understand what the advantage of that is.

**Comment by ambrop72** [comment]
@ttuegel I would just be careful to have the same exact derivation, so that we don't have a (default) kde4.kdelibs and kdeApps.kdelibs which are the same version but different. Passing the URL achieves this automatically, but I agree it's not as flexible as passing a derivation.

If we are to pass a derivation, then would it make sense to kill the kde4.kdelibs derivation, and make kde4 by default just use kdeApps_stable.kdelibs?

**Comment by ambrop72** [comment]
But, I do suspect there is a difference between the kde4 and kdeApps expressions for kdelibs, which is the underlying cause for the style problem I'm seeing. Could be that getting everything to use kdeApps.kdelibs will just make it broken for all apps. So from this perspective I prefer reusing the kde4 derivation. But I suppose we could always debug and fix it.

**Comment by ttuegel** [comment]
I don't think the expressions are different. I copied the expression out of `kde4` to make `kdeApps`. I have my system rebuilt now so that there is only one version of `kdelibs` around, and I don't have any theme problems. I'm closing this for now (I've merged the `kdelibs` update for `kde4`), but please re-open it if you continue to have problems.

**Comment by ambrop72** [comment]
I've rebuilt my system with your change. The result is that now all the KDE4 apps are having style problems. See picture: http://www.pasteall.org/pic/show.php?id=85653 . Here Konsole is a KF5 app and it has the white Breeze theme, but KDevelop is grey. The KDE5 application style settings have no effect on KDE4 apps now - previously, the ones that got linked to kde4.kdelibs were integrating well.

I think the reason for the change is the default profile path. The kdeApps.kdelibs used ~/.kde5, while the old kde4.kdelibs used ~/.kde . You can see a proof of this in my screenshot, where KDevelop files are put into ~/.kde5, and KDevelop has "forgotten" about my previous configuration due to this change.

**Comment by ambrop72** [comment]
Of course I consider it inappropriate to change the profile path for all KDE4 apps. People have settings in .kde and nobody expects it to be lost. So I would change the kdelibs expression to force the profile path to ~/.kde.

Note: I can't reopen this issue since I didn't close it.

**Comment by ttuegel** [comment]
Well, that CMake flag was not supposed to change KDEHOME for KDE 4 applications. `$HOME/.kde5` is supposed to be used for KDE 5 applications, and `$HOME/.kde` for KDE 4 applications. I did this intentionally so that the KDE 5 configuration wouldn't clobber existing KDE 4 things. But now that I look at it, KDE 5 applications are using the XDG-compliant `$HOME/.config` path. I suppose that means it's safe to use `$HOME/.kde` for KDEHOME everywhere.

Thanks for tracking that down! Sorry, I didn't realize GitHub wouldn't let you re-open closed issues. That seems strange.

**Comment by ambrop72** [comment]
Couldn't we just keep with what used to be - .kde for KDE4 and .kde5 for KF5? It worked fine for me. All that needs to be done is to build kdelibs with KDEHOME .kde.

**Comment by ttuegel** [comment]
@ambrop72 Well, that's what I'm saying: none of the KF5 apps are using `.kde5`! They're all putting their settings in `$HOME/.config`. Looking in my own `$HOME/.kde5`, it only contains settings for applications that are still using `kdelibs-4`, i.e. that haven't been ported to Frameworks yet.

**Comment by ambrop72** [comment]
Ah, yes, misunderstanding. I just deleted .kde* and loaded up the desktop with my original fix that removes kdeApps.kdelibs, and indeed only .kde was created, no kde5.

**Comment by ttuegel** [comment]
Fixed in master.
